compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala

```diff
@@ -847,10 +872,10 @@ abstract class PrepJSInterop extends plugins.PluginComponent
 
         // Traits must be pure interfaces
         if (sym.owner.isTrait && sym.isTerm && !sym.isConstructor) {
-          if (!sym.isDeferred) {
+          if (!sym.isDeferred && !sym.hasAnnotation(JSOptionalAnnotation)) {
             reporter.error(tree.pos,
-                "A Scala.js-defined JS trait can only contain abstract members")
-          } else if (isPrivateMaybeWithin(sym)) {
+                "A Scala.js-defined JS trait can only contain abstract members and/or @JSOptional members")
+          } else if (sym.isMethod && isPrivateMaybeWithin(sym)) {
```

Comment:
Because now you can have a private field created by a concrete `val` or `var`. Previously we only had abstract vals and vars, for which `typer` does not create a private field.